### DISCUSSION TITLE
Improve Web Search Intelligence in LLM Responses

### DISCUSSION THREAD

**Discussion by odefta** [open-webui/open-webui#4648] [General]
## Current Behavior
When web search is enabled and a user asks a specific question, the system performs a **literal search of the entire question**. For example:

Query: "I don't know how to modify the column type of a materialized view in Oracle 19c from char(5) to varchar2(100)... please help me"

Current search: [entire user query]

This often leads to **poor or irrelevant search results**, as the search query is too specific and unlikely to match existing web content directly.

## Proposed Behavior
1. The LLM should **first analyze the user's query** to determine the core information need.
2. Based on this analysis, it should **generate a more generalized and effective search query**.
3. It should then **perform the web search using this optimized query**.
4. Finally, it should **use the search results to inform its response** to the user's original question.

### Example:
User Query: "I don't know how to modify the column type of a materialized view in Oracle 19c from char(5) to varchar2(100)... please help me"

- Current Search: [entire user query]
- **Proposed Search: "alter materialized view column type oracle 19c"**

The LLM would then use the results from this more targeted search to construct a detailed response to the user's specific question.

## Benefits
- **More relevant and useful web search results**
- **Improved ability of the LLM to provide accurate and helpful responses**
- Better user experience, especially for technical or specific queries
- Closer alignment with advanced AI platforms like Abacus.AI

## Technical Considerations
- This feature will require the LLM to perform **two-step processing**: 
  1. Analyze the query and generate a search term
  2. Use the search results in formulating a response
- The system may need to implement a form of **query understanding or intent classification** to generate effective search terms.
- Consider implementing this as an **optional feature** that can be toggled on/off, as some users might prefer the current literal search in certain scenarios.

---

This enhancement would significantly improve the practical utility of the web search feature in OpenWebUI, especially for technical and specific queries (where targeted information is crucial).

**Comment by justinh-rahb** [comment]
<img width="662" alt="Screenshot 2024-08-06 at 1 19 01 AM" src="https://github.com/user-attachments/assets/644bb801-2cef-4d42-9aab-d8b6238a4432">

**Comment by odefta** [comment]
Thanks, it works after changing the search query prompt and lowering the prompt length.
(I didn't know that I can customize that)

I've used something like:
Today is {{CURRENT_DATE}}.
Extract key concepts from the query and create a short, focused query (prioritizing technical terms in max 10 words).

Example:
Query: 'How do I modify the column type of a materialized view in Oracle 19c from char(5) to varchar2(100)? Please help me!'
Response: 'alter materialized view column type Oracle 19c char varchar2'

Output only the keywords.

Question:
{{prompt:end:4000}}
